### ISSUE TITLE
Always 0%

### ISSUE THREAD

**Issue by Streel88** [nakiostudio/xcov#22]
Everyone I run xcov I am getting 0%
- I have **code coverage** enabled in the Test Config
- Ensured the Scheme **shared** 

I created a new Project with some quick code and Xcode is showing the count beside each function inside the IDE.

My Command
`xcov -s AutoTest -p ./AutoTest.xcodeproj -o ./build/reports/xcov_output

I get the following Output

> [14:47:22]: xcrun xcodebuild -list -project './AutoTest.xcodeproj'
> [14:47:23]: Skipping file blacklisting as no ignore file was found at path ./.xcovignore
> 
> +-----------------------------+------------------------------------------+
> |                          Summary for xcov 0.7                          |
> +-----------------------------+------------------------------------------+
> | scheme                      | AutoTest                                 |
> | project                     | ./AutoTest.xcodeproj                     |
> | output_directory            | ./build/reports/xcov_output              |
> | derived_data_path           | /Users/christopher.streel/Desktop/Build/ |
> | minimum_coverage_percentage | 0.0                                      |
> | ignore_file_path            | ./.xcovignore                            |
> | include_test_targets        | false                                    |
> | skip_slack                  | false                                    |
> +-----------------------------+------------------------------------------+
> 
> [14:47:23]: $ /Library/Ruby/Gems/2.0.0/gems/xcov-0.7/lib/xcov-core/bin/xcov-core -s /Users/christopher.streel/Desktop/Build/Logs/Test/42D0D29B-E7A5-4A7D-B5BF-36058F0AE62E.xccoverage -o /var/folders/wz/p18ghh6561v49_s7mgdsvhg1lyr51y/T/report.json20160419-48755-1yu5l3q
> [14:47:23]: ▸ Loading...
> [14:47:23]: ▸ ------ xCov-Core ------
> [14:47:23]: ▸ Opening .xccoverage file at path: /Users/christopher.streel/Desktop/Build/Logs/Test/42D0D29B-E7A5-4A7D-B5BF-36058F0AE62E.xccoverage
> [14:47:23]: ▸ Parsing .xccoverage file...
> [14:47:23]: ▸ File successfully parsed
> [14:47:23]: ▸ Serializing coverage report...
> [14:47:23]: ▸ Report successfully serialized
> [14:47:23]: ▸ Writing report on disk...
> [14:47:23]: ▸ Coverage report successfully created at path: /var/folders/wz/p18ghh6561v49_s7mgdsvhg1lyr51y/T/report.json20160419-48755-1yu5l3q
> +--------------+-----------+
> |   xcov Coverage Report   |
> +--------------+-----------+
> | AutoTest.app | 0%        |
> +--------------+-----------+

I've attached a screenshot, if you need any more info let me know.

I love the html output, just hope the Cover Coverage gets corrected

<img width="2085" alt="screen shot 2016-04-19 at 2 53 12 pm" src="https://cloud.githubusercontent.com/assets/2611041/14628143/914345d2-063e-11e6-8fe6-5c61b3de9435.png">

**Comment by Streel88** [comment]
Oh, I am also running Xcode 7.3

**Comment by Streel88** [comment]
So I found a difference between using xcodebuild and the IDE

With the IDE it works but with xcodebuild it doesn't

**Comment by nakiostudio** [comment]
Hi @Streel88 👋 

Try adding the `build` argument to your `xcodebuild` command. That should be the only difference between running the tests from Xcode and running them from the console. I hope it helps!

**Comment by Streel88** [comment]
Hi @nakiostudio 

I tried that and still no luck, to see if you can replicate my issue

1. Purge your Derived Data, either manually or via Xcode
2. Execute the following command

**Comment by nakiostudio** [comment]
Hi again @Streel88 👋 

You are absolutely right, the `Debug Executable` flag must be enabled in order to get the coverage information. Could you please try adding the `-configuration "Debug"` argument to your command?

**Comment by Streel88** [comment]
Still no luck all that does is change the build configuration, from what I've learned about it is one of the executables it produces under ../CodeCoverage/.. needs to be executed which it doesn't seem to do with xcodebuild.

I will keep looking into it further to see if I can't figure out the correct command

**Comment by nakiostudio** [comment]
Honestly, I have no idea what's wrong... if you can publish a sample project with this problem so I can clone and test it. I haven't been able to repro this on my machine.

**Comment by Streel88** [comment]
Sorry for the delayed reply, here is the dummy Project I put together to test
[AutoTest.zip](https://github.com/nakiostudio/xcov/files/237788/AutoTest.zip)

Scenarios
1. Test with IDE, code completion enabled. Run xcov -> Success
2. Test with IDE, code completion enabled, Debug Disabled. Run xcov -> Fail (But expected since IDE also shows no coverage)
2. Clean/Delete DerivedData Folder, Test with xcodebuild, Run xcov -> Fail

Commands
xcodebuild
`xcodebuild build test -project ./AutoTest.xcodeproj/ -scheme AutoTest -destination 'platform=iOS Simulator,name=iPhone 5s,OS=9.3' -enableCodeCoverage YES`
xcov
`xcov -s AutoTest -p ./AutoTest.xcodeproj/ -o ./xcov_output`

**Comment by redok016** [comment]
I have the same issue. Is there any updated information related with this?

**Comment by nakiostudio** [comment]
Hi @redok016 :wave:

I'm looking at the problem, I'll try to get a fix out soon.

Thanks @Streel88 for the additional information 👍

**Comment by mathcarignani** [comment]
I have the same issue with my UI tests, when I run `scan` (use xcodebuild) I don't have the coverage from those tests

**Comment by triage** [comment]
I had this issue till recently. We have an objective-c project that we've only recently started implementing swift too. All of my tests would run, but it appeared that only swift files were used in the computation of coverage reports -- leading to percentages like 7%.

Turned on this build setting and now my objective-c files are included in coverage reports properly ... which was not a part of the readme instructions for xcov:
<img width="402" alt="screen shot 2016-06-15 at 7 17 06 am" src="https://cloud.githubusercontent.com/assets/160455/16083257/33f90008-32c9-11e6-9867-1c89952ce184.png">

**Comment by thelvis4** [comment]
After some investigations, here is what I've noticed:
* Running tests from Xcode IDE differs from running tests using `xcodebuild`:

Comparing `Logs/Test/XXX/Session-<some_uid>.log`:

**Xcode**: 
```
Launching with Xcode.DebuggerFoundation.Launcher.LLDB
...
"wait_for_debugger" = 1;
```
**xcodebuild**:
```
Launching with Xcode.IDEFoundation.Launcher.PosixSpawn
...
 "wait_for_debugger" = 0;
```

* `LauncherIdentifier` depends on `Debug Executable` flag (in Xcode scheme):

**`Debug Executable` enabled**: 
```
selectedDebuggerIdentifier = "Xcode.DebuggerFoundation.Debugger.LLDB"
selectedLauncherIdentifier = "Xcode.DebuggerFoundation.Launcher.LLDB"
```
**`Debug Executable` disabled**: 
```
selectedDebuggerIdentifier = ""
selectedLauncherIdentifier = "Xcode.IDEFoundation.Launcher.PosixSpawn"
```
So, when running tests using `xcodebuild`, it behaves as when `Debug Executable` is disabled, no matter what the actual state of  `Debug Executable` flag is.

* When `Debug Executable` flag is disabled, Unit tests give corect coverage data while Unit Tests **do not**.

Running Unit tests with `Xcode.IDEFoundation.Launcher.PosixSpawn` Launcher Identifier (similar with having  `Debug Executable` flag disabled) doesn't affect coverage data. Running UI Tests with these configurations, gives invalid coverage data (0% for tested target).
I suppose the differences comes from different mechanisms used for starting UI Tests and Unit Tests.
[Screenshot from a WWDC '16 presentation](https://s32.postimg.org/msd8951jp/Screen_Shot_2016_07_12_at_1_42_24_PM.png)
[WWDC '16 `Advanced Testing and Continuous Integration` video](https://developer.apple.com/videos/play/wwdc2016/409/)

**Comment by nakiostudio** [comment]
Great stuff @thelvis4 👍 I've been digging a bit into it and there is very little information. I will test a couple of things during the weekend though.

Thanks!

**Comment by redok016** [comment]
@thelvis4,@nakiostudio thanks for your great effort

**Comment by raven** [comment]
@thelvis4 have you filed a radar? Just wanting to dupe it if you have one already :) 
Also, have you had a chance to check if the behaviour is different with Xcode 8?

**Comment by thelvis4** [comment]
@raven Sorry, haven't had the time to further investigate (with Xcode 8) and to file a radar. But I would be glad to dupe the radar if you could file one. I believe we should check this on Xcode 8, as I doubt they will release another version of Xcode 7.

**Comment by stevepeak** [comment]
At [Codecov](https://codecov.io) we have resolved this issue with the help of customers. It appears that `xcpretty` corrupted the coverage reports. https://github.com/codecov/example-swift/issues/14#issuecomment-249427948

Hope this is related and find it useful. Thank you!

**Comment by nakiostudio** [comment]
Hi @stevepeak 👋 

Thanks for the heads up, however the issue we are facing is pretty different. xcov processes swift 3 coverage reports without any problems. In addition, xcov works directly with `.xccoverage` files therefore it has nothing to do with `xcpretty` 😞 

Thank you once again and keep up the good work with Codecov!

**Comment by iwllyu** [comment]
I created a project that demonstrates this bug

https://github.com/iwllyu/UITestCoverageBug

If you run the tests via XCode IDE you get 80% of ViewController.swift coverage. If you run the tests via xcodebuild CLI you only get 40% coverage for ViewController.swift.

**Comment by thelvis4** [comment]
@iwllyu Thank you for submitting the radar and putting it on open radar as well. 👍 
I'm going to duplicate it.
btw, we use "dupe" as a short form of "duplicate".

**Comment by nakiostudio** [comment]
Thank you @iwllyu! I will take a look this weekend. Hopefully we will come up with a final solution for this issue.

**Comment by EliseLeb** [comment]
Has a solution been found?

**Comment by iwllyu** [comment]
I got this on on 12/16/2016 @ 3:17AM

> 
> Engineering has determined that your bug report (29033348) is a duplicate of another issue (28900146) and will be closed.
> 
> The open or closed status of the original bug report your issue was duplicated to appears in the yellow "Duplicate of XXXXXXXX" section of the bug reporter user interface. This section appears near the top of the right column's bug detail view just under the bug number, title, state, product and rank.
> 
> An example of the duplicate section from the bug reporter user interface with your bug and the duplicate bug info is included below:
> 
> 29033348 Test Coverage reports from xcodebuild CLI ignore coverage from UI tests
>                   
> State: Closed                   Product:
> Rank: No Value
> ----------------------------------------------------------------------------------------------
> Duplicate of 28900146 (Open or Closed; log in to see the actual state)
> ----------------------------------------------------------------------------------------------
>

**Comment by michaellady** [comment]
@iwllyu thank you for helping me realize I'm still sane!

**Comment by thelvis4** [comment]
@EliseLeb This is definitely an inconsistency (bug) in Apple's tools. And it's still reproducible with Xcode 8.2.

Xcode IDE and `xcodebuild` use different UITests launchers (and debuggers):
* Xcode IDE  uses this plugin `Xcode.app/Contents/PlugIns/DebuggerLLDB.ideplugin`
whereas
* `xcodebuild` uses `IDEFoundation` framework `Xcode.app/Contents/Frameworks/IDEFoundation.framework` (and does't attach any debugger)

I couldn't make (force) `xcodebuild` to take `Xcode.DebuggerFoundation.Launcher.LLDB` as a launcher, although I tried half of a day in different undocumented (and probably stupid) ways. Probably it's an implementation detail and cannot be influenced externally.

At this point, I will wait for Apple to fix this bug/inconsistency.  BTW, my radar was also marked as `Duplicate of 28900146 (Open)`.

**Comment by EliseLeb** [comment]
@thelvis4 Thanks for the update!  I also filed a bug with apple but not holding my breath.  What a bummer for UI tests.

**Comment by earltedly** [comment]
I've filed a radar too - thanks everyone in this thread for saving me a few extra hours working out what was going on! Has anyone tried in the latest beta yet?

**Comment by thelvis4** [comment]
@neonichu told me that the ticket is under investigation. 🤞

**Comment by earltedly** [comment]
@thelvis4 @neonichu it also might relate to <rdar://problem/28571841> which I filed about UI tests always reporting 0% coverage for linked frameworks. There's a simple example project attached.

**Comment by michaellady** [comment]
Anyone try this on 8.3.2 yet?

**Comment by iwllyu** [comment]
@michaellady you're welcome to clone the project here https://github.com/iwllyu/UITestCoverageBug and try it out on 8.3.2

**Comment by michaellady** [comment]
Hi @iwllyu, I'm still seeing the command line issue on 8.3.2

**Comment by iwllyu** [comment]
thanks, I've updated the readme in that project

**Comment by JohnCoates** [comment]
I looked into fixing this. Found the difference in logs that @thelvis4 did. Figured I'd see if there was an undocumented flag I could pass to xcodebuild to use the LLDB launcher instead of PosixSpawn. Decompiling didn't find anything, so I decided to force it myself to see if it would work. Modifying xcode build to use LLDB instead of PosixSpawn indeed works, but its fragility won't be an ideal solution to implement into most projects. Most of us are probably stuck waiting for Apple to implement a fix, so file a Radar!

Proof of Concept commands:
````
# we'll run xcodebuild through a debugger so we can alter its execution
lldb

# set our debugging target
target create /Applications/Xcode.app/Contents/Developer/usr/bin/xcodebuild

# launch process with arguments and stop at entry point
process launch -s -- -sdk iphonesimulator -destination 'name=iPhone SE' test

# set a breakpoint on an argument where we can alter the process launcher & debugger
# the method won't be loaded into the process yet, so don't worry about the warning lldb gives, the breakpoint will be resolved automatically as soon as the method is available
breakpoint set -F "+[IDELaunchParametersSnapshot launchParametersWithSchemeIdentifier:launcherIdentifier:debuggerIdentifier:launchStyle:runnableLocation:debugProcessAsUID:workingDirectory:commandLineArgs:environmentVariables:architecture:platformIdentifier:buildConfiguration:buildableProduct:deviceAppDataPackage:allowLocationSimulation:locationScenarioReference:showNonLocalizedStrings:language:region:routingCoverageFileReference:enableGPUFrameCaptureMode:enableGPUValidationMode:debugXPCServices:debugAppExtensions:internalIOSLaunchStyle:internalIOSSubstitutionApp:launchAutomaticallySubstyle:]"

# now let's add some commands to run when the breakpoint hits
# we'll replace the launcher & debugger arguments passed to the function and then continue execution
break command add
po $rcx = (unsigned long)IDEDefaultLauncherIdentifier
po $r8 = (unsigned long)IDEDefaultDebuggerIdentifier
continue
DONE

# resume execution from the entry point. coverage should be correct now
continue
````
Proof of concept code for injection as a library: https://gist.github.com/JohnCoates/605319ab31c0aa05c77973481aeae4b2

**Comment by iwllyu** [comment]
https://github.com/coryalder is reporting that this is fixed in XCode9

https://github.com/iwllyu/UITestCoverageBug/issues/4

**Comment by michaellady** [comment]
Woot!

**Comment by nakiostudio** [comment]
Amazing work by the dev tools team in this release 👏

**Comment by asos-petergoldsmith** [comment]
*OMG*

**Comment by iwllyu** [comment]
Oops we may have jumped the gun. 

https://github.com/iwllyu/UITestCoverageBug/issues/4#issuecomment-317430848

tl;dr: it looks like both xcode and xcodebuild are not covering UI tests. xcode could have erroneously reporting coverage before and  Apple intends for UI tests not to be part of code coverage.

Well, at least they're consistent now

**Comment by Vkt0r** [comment]
@iwllyu Any update in this?? I'm running Xcode 9 and after ran all the tests the UI Tests are missed from the xcoverage report. Xcov is only reporting the unit tests.

**Comment by vgorloff** [comment]
Xcode Version 9.3 beta (9Q98q) seems resolves issue with missed UI Tests coverage as long as `Debug executable` not enabled in Test schema. Details: https://stackoverflow.com/a/48477190/1418981

**Comment by raven** [comment]
> Xcode Version 9.3 beta (9Q98q) seems resolves issue with missed UI Tests coverage as long as Debug executable not enabled in Test schema. Details: https://stackoverflow.com/a/48477190/1418981

I've verified this is indeed working again 🎉

**Comment by stephen-oconnor-cp** [comment]
Sorry to comment on this after so long, but this is the only issue that relates to the problem I'm having. 

Not even unit tests are running right now.  If I run them in Xcode, it generates coverage and does everything as expected.  If it's run via run_tests / scan, it doesn't appear that unit tests are even running, even though it has output like this:

```
[10:13:55]: Running Tests: ▸ Touching CameraDummyTests.xctest (in target 'CameraDummyTests' from project 'CameraDummy')
[10:13:55]: ▸ Processing empty-BuildPipeSandbox.plist
[10:13:55]: ▸ Processing empty-BuildPipeSandboxTests.plist
[10:13:56]: ▸ Linking BuildPipeSandboxTests
[10:13:56]: Running Tests: ▸ Touching BuildPipeSandboxTests.xctest (in target 'BuildPipeSandboxTests' from project 'BuildPipeSandbox')
[10:13:56]: Running Tests: ▸ Touching BuildPipeSandbox.app (in target 'BuildPipeSandbox' from project 'BuildPipeSandbox')
[10:13:56]: ▸ Linking BuildPipeSandboxUITests
[10:13:56]: ▸ Processing empty-BuildPipeSandboxUITests.plist
[10:13:56]: ▸ Copying Info.plist
[10:13:56]: Running Tests: ▸ Touching BuildPipeSandboxUITests.xctest (in target 'BuildPipeSandboxUITests' from project 'BuildPipeSandbox')
[10:13:56]: ▸ Test build Succeeded
[10:13:56]: Copying .xctestrun file
```

Can anybody help with this?

**Comment by michaellady** [comment]
This sounds like you should open a new issue if this was closed that long
ago. I don’t even do mobile devops anymore, lol.
